Skip Triton templates in MM max autotune with zero-size inputs

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106865

Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
@torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Additionally, early exit conditions have been added to the mm and mm_plus_mm Triton templates on `M * N * K == 0`, to prevent issues when autotuning was done on non-zero-size inputs with dynamic shapes, then zero-size inputs are encountered by the compiled model.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: @eellison

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

aakhundov committed:
Fix MM max autotune for zero-size inputs

Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
@torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: @eellison

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ipiszy commented:
Wonder what's the exact error from Triton?

aakhundov commented:
> Wonder what's the exact error from Triton?

Do you know a way to recover the Triton error? My impression of the exception is that it's left behind the scenes. But I'm happy to get it and share if you know how.

ipiszy commented:
I see, Triton just exits abnormally. Maybe it's caused by division by zero (rm % M).
It seems to me that we should just update the Triton mm template to exit early when input shape is zero? In this case it can also handle dynamic shapes with zero inputs.

aakhundov commented:
> It seems to me that we should just update the Triton mm template to exit early when input shape is zero?

I think, this is a good idea. Although, should we still skip the Triton templates if the inputs used during compilation are zero-size? Because the result of this autotuning would not be very helpful.

eellison commented:
Interestingly, it's only https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm_common.py#L61 that will fail, all the other configs work.

aakhundov committed:
Update on "Skip Triton templates in MM max autotune with zero-size inputs"


Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: eellison

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

jansel commented:
Another option here would be to add a decomp for mm/bmm similar to: https://github.com/pytorch/pytorch/pull/106593/

Where if any of the sizes are zero, we just decompose to `torch.empty([])` and remove the mm from the graph.

eellison commented on torch/_inductor/kernel/mm_plus_mm.py:
```diff
@@ -26,6 +27,9 @@
     M = {{size("A", 0)}}
     N = {{size("B", 1)}}
     K1 = {{size("A", 1)}}
+    if M * N * K1 == 0:
```
this change isn't being tested because of the other changes.. what happens when you remove the other changes, and just include this ?

aakhundov replied:
I've tried this. The generated Triton kernels are running fine and take exactly the same time as the aten one.

aakhundov commented:
@pytorchbot merge

aakhundov commented:
@pytorchbot label "topic: not user facing"

aakhundov commented:
@pytorchbot merge

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -29,6 +30,9 @@
     K = {{size("A", 1)}}
+    if M * N * K == 0:
+        # early exit due to zero-size input(s)
+        return
```
I think we also need to handle the suffix

aakhundov replied:
Great catch! Looking at the template code, I believe, we can simply omit `* K` from the condition.

aakhundov committed:
Update on "Skip Triton templates in MM max autotune with zero-size inputs"


Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Additionally, early exit conditions have been added to the mm and mm_plus_mm Triton templates on `M * N * K == 0`, to prevent issues when autotuning was done on non-zero-size inputs with dynamic shapes, then zero-size inputs are encountered by the compiled model.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: eellison

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

aakhundov committed:
Update on "Skip Triton templates in MM max autotune with zero-size inputs"


Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Additionally, early exit conditions have been added to the mm and mm_plus_mm Triton templates on `M * N * K == 0`, to prevent issues when autotuning was done on non-zero-size inputs with dynamic shapes, then zero-size inputs are encountered by the compiled model.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: eellison

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

aakhundov commented:
Removed `* k` and `* K` from the lowering (autotuning) functions and the early exit conditions in the Triton templates, respectively. This way, computation of all-zero output on `K=0` (possibly, with a fused epilogue) will happen as expected. Also, it seems to make sense to autotune in the `k=0` case, as Triton can outperform ATen operators in such cases.

aakhundov committed:
Update on "Skip Triton templates in MM max autotune with zero-size inputs"


Summary:

MM max autotune (and friends) crash when one of the inputs is zero-size.

E.g., running this code:

```
torch.compile()
def fn(x, y):
    return torch.mm(x, y)

inps = [torch.rand([0, 30]), torch.rand([30, 40])]
inps = [x.to(device="cuda") for x in inps]
out = fn(*inps)
```

with this command:

```
TORCHINDUCTOR_MAX_AUTOTUNE=1 python test.py
```

raises this error (the top of the stack trace omitted for brevity):

```
...
  File "/data/users/aakhundov/pytorch/torch/_inductor/kernel/mm.py", line 119, in tuned_mm
    return autotune_select_algorithm("mm", choices, [mat1, mat2], layout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 960, in autotune_select_algorithm
    return _ALGORITHM_SELECTOR_CACHE(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 787, in __call__
    timings = self.lookup(
              ^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/codecache.py", line 267, in lookup
    timings[choice] = benchmark(choice)
                      ^^^^^^^^^^^^^^^^^
  File "/data/users/aakhundov/pytorch/torch/_inductor/select_algorithm.py", line 774, in autotune
    raise ErrorFromChoice(msg, choice, benchmark_fn.debug_str())
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
LoweringException: ErrorFromChoice: Please run `ptxas /tmp/compile-ptx-src-bfb1c6` to confirm that this is a bug in `ptxas`

From choice TritonTemplateCaller(/tmp/torchinductor_aakhundov/z7/cz7n7nn6rdlaelu4pbaaurgmu74ikl6g76lkngwawrevlfxlc6re.py, ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=32, BLOCK_M=16, BLOCK_N=64, EVEN_K=False, GROUP_M=8, num_stages=2, num_warps=4)
inputs = [
    torch.empty_strided((0, 30), (30, 1), dtype=torch.float32, device='cuda'),
    torch.empty_strided((30, 40), (40, 1), dtype=torch.float32, device='cuda'),
]
out = torch.empty_strided((0, 40), (40, 1), dtype=torch.float32, device='cuda')

  target: aten.mm.default
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='arg1_1', layout=FixedLayout('cuda', torch.float32, size=[0, s0], stride=[s0, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='arg3_1', layout=FixedLayout('cuda', torch.float32, size=[s0, s1], stride=[s1, 1]))
  ))
```

This PR adds a check to skip Triton templates in the `mm`, `addmm`, `mm_plus_mm` autotuning when the product of the MM problem shape (`m * n * k`) is zero.

Additionally, early exit conditions have been added to the mm and mm_plus_mm Triton templates on `M * N * K == 0`, to prevent issues when autotuning was done on non-zero-size inputs with dynamic shapes, then zero-size inputs are encountered by the compiled model.

Test Plan:

```
$ python test/inductor/test_max_autotune.py -v

...

----------------------------------------------------------------------
Ran 16 tests in 29.569s

OK
```

Reviewers: eellison

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

aakhundov commented:
@pytorchbot merge

aakhundov commented:
@pytorchbot merge -f "failing CI jobs look unrelated; trunk is broken"
